packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2658,17 +2672,41 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
             //    failure (batch is too large). Pushing them earlier and outside of main batch should alleviate
             //    these issues.
             // Cons:
-            //    With large batches, relay service may throttle clients. Clients may disconnect while throttled.
+            //    1. With large batches, relay service may throttle clients. Clients may disconnect while throttled.
             //    This change creates new possibility of a lot of newly created data stores never being referenced
             //    because client died before it had a change to submit the rest of the ops. This will create more
             //    garbage that needs to be collected leveraging GC (Garbage Collection) feature.
+            //    2. Sending ops out of order means they are excluded from rollback functionality. This is not an issue
+            //    today as rollback can't undo creation of data store. To some extent not sending them is a bigger
+            //    issue than sending.
             // Please note that this does not change file format, so it can be disabled in the future if this
             // optimization no longer makes sense (for example, batch compression may make it less appealing).
             if (this._flushMode === FlushMode.TurnBased && type === ContainerMessageType.Attach &&
                     this.mc.config.getBoolean("Fluid.ContainerRuntime.disableAttachOpReorder") !== true) {
-                this.flushBatch([message]);
+                if (!this.pendingAttachBatch.push(message)) {
+                    this.flushBatch(this.pendingAttachBatch.popBatch());
+                    if (!this.pendingAttachBatch.push(message)) {
+                        throw new GenericError(
+                            "BatchTooLarge",
+                            /* error */ undefined,
+                            {
+                                opSize: message.contents.length,
+                                count: this.pendingAttachBatch.length,
+                                limit: this.pendingAttachBatch.limit,
```

Comment:
this would be misleading if the batch manager blocks because of the soft limit, as this getter always returns the hard limit. Perhaps it needs to return the actual limit it uses.

Reply:
The behavior (see UTs & implementation) is that soft limit is only engaged when the queue is not empty. Thus, we flush the queue on first failure, ensuring that soft limit is not engaged on attempt to push that message into the queue. I'll add comments in this flow.